src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1906,204 @@ MutableBigInteger sqrt() {
+                            ONE, needRemainder ? new MutableBigInteger((int) (x - 1)) : null
+                    };
+
+                long s = (long) Math.sqrt(x); // Math.sqrt is exact
+                return new MutableBigInteger[] {
+                        new MutableBigInteger((int) s),
+                        needRemainder ? new MutableBigInteger((int) (x - s * s)) : null
+                };
+            }
+
+            // Initial estimate is the square root of the unsigned long value.
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
+            if (s > LONG_MASK // avoid overflow of s * s
+                    || Long.compareUnsigned(x, s * s) < 0) {
+                s--;
+            }
+
+            return new MutableBigInteger[] {
+                    new MutableBigInteger((int) s),
+                    needRemainder ? new MutableBigInteger(x - s * s) : null
+            };
+        }
+
+        long bitLength = this.bitLength();
+        if (bitLength != (int) bitLength)
+            throw new ArithmeticException("bitLength() integer overflow");
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.primitiveLeftShift(shift);
+        }
+
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder && !sqrtRem[1].isZero()) {
+                final int s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0) { // An optimization
+                    // s0 is at most 15 bit long
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul(s0 << 1, doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
+                }
+                sqrtRem[1].primitiveRightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
+
+    /**
+     * Assume {@code 2 <= len <= intLen && Integer.numberOfLeadingZeros(value[offset]) <= 1}
+     * @implNote The implementation is based on Zimmermann's works available
+     * <a href="https://inria.hal.science/inria-00072854/en/">  here</a> and
+     * <a href="https://www.researchgate.net/publication/220532560_A_proof_of_GMP_square_root">  here</a>
+     */
+    private MutableBigInteger[] sqrtRemZimmermann(int len, boolean needRemainder) {
+        if (len == 2) { // Base case
+            long x = ((value[offset] & LONG_MASK) << 32) | (value[offset + 1] & LONG_MASK);
+            long s = (long) Math.sqrt(x > 0 ? x : x + 0x1p64);
```

Comment:
Again, this needs a proof that doing so is always correct.

```suggestion
            long s = (long) Math.sqrt(x >= 0 ? x : x + 0x1p64);
```

Reply:
Sure, but again: changing `>` to `>=` makes the code correct even for `0`.